### PR TITLE
feat(Web, Home): API Integration

### LINKED ISSUE #1: handle crash on homepage
client is "expecting" JSON data, but when it's empty, client will crash the app. The interesting part is this code:

```reason
React.useEffect0(() => {
    Js.Promise.(
      fetch("/api/url/list")
      |> then_(response => response##json())
      |> then_(responseJSON => {
           setState(_previousState => URLs(responseJSON));
           Js.Promise.resolve();
         })
      |> catch(_err => {
           setState(_previousState => ErrorFetch);
           Js.Promise.resolve();
         })
      |> ignore
    );

    None;
  });
}
```

On `stats` fragments, I trick that by rejecting the promise using 204 status code:

```go
func GetHitsStatsBySlug(db *sql.DB, w http.ResponseWriter, r *http.Request) {
	params := r.URL.Query()
	slug := params.Get(":slug")
	payload := models.GetHitsBySlug(db, slug)

	if payload == nil {
		emptyData := make([]string, 0)

		// FIXME: this is temporary fix to our frontend!
		ResponseJSON(w, http.StatusNoContent, notFound{emptyData})

		return
	}

	ResponseJSON(w, http.StatusOK, payload)
}
```

Absolutely this is not a good solution, I need to know how to handle "unexpected" response in Reason, so if the payload I need is like this:

```json
{
  "data": [{ "id": 124, "url": "https://evlfctry.pro/send-nudes" }]
}
```

But the given response is like this:

```json
{
  "data": null
}
```

How to handle that beautifully in Reason so the client will not crash?

### PR DESCRIPTION
**Pull request by r17x**
- install & setup relude with `npm install decco relude{,-fetch,-reason-react} bs-{fetch,abstract}`
- add src/API.re
- implement API.re in src/fragments/Home.re
- fix https://github.com/evilfactorylabs/gow/issues/1

### CODE REVIEW

**Comment by faultables** [comment]
There are [errors](https://hub.docker.com/repository/registry-1.docker.io/evilfactorylabs/gow/builds/40c950ea-6ae9-49fe-abd6-b89768c4e337) related with the dependencies you recently added.

```
FAILED: src/Decco_Codecs.reast
/app/node_modules/bs-platform/lib/bsc.exe -w -30-40+6+7+27+32..39+44+45+101+A-9-40-42 -color always -ppx /app/node_modules/decco/./ppx -bs-quiet -bs-super-errors -o src/Decco_Codecs.reast -bs-syntax-only -bs-binary-ast /app/node_modules/decco/src/Decco_Codecs.re
/app/node_modules/decco/./ppx: line 14: /app/node_modules/decco/./ppx-linux.exe: not found
We've found a bug for you!
/app/node_modules/decco/src/Decco_Codecs.re
Error while running external preprocessor
Command line: /app/node_modules/decco/./ppx '/tmp/camlppx09a53f' '/tmp/camlppx6a3356'
[2/6] [34mBuilding[39m [2msrc/Decco.reast[22m
FAILED: src/Decco.reast
/app/node_modules/bs-platform/lib/bsc.exe -w -30-40+6+7+27+32..39+44+45+101+A-9-40-42 -color always -ppx /app/node_modules/decco/./ppx -bs-quiet -bs-super-errors -o src/Decco.reast -bs-syntax-only -bs-binary-ast /app/node_modules/decco/src/Decco.re
/app/node_modules/decco/./ppx: line 14: /app/node_modules/decco/./ppx-linux.exe: not found
We've found a bug for you!
/app/node_modules/decco/src/Decco.re
Error while running external preprocessor
Command line: /app/node_modules/decco/./ppx '/tmp/camlppx03369d' '/tmp/camlppx862ef2'
FAILED: subcommand failed.
Failure: /app/node_modules/bs-platform/lib/ninja.exe
Location: /app/node_modules/decco/lib/bs

npm
 
ERR!
 
code
 ELIFECYCLE

npm
 
ERR!
 
errno
 2

npm
 
ERR!
 @evilfactorylabs/gow@0.2.0 build: `bsb -make-world && webpack`

npm
 
ERR!
 Exit status 2

npm
 
ERR!


npm
 
ERR!
 Failed at the @evilfactorylabs/gow@0.2.0 build script.

npm
 
ERR!
 This is probably not a problem with npm. There is likely additional logging output above.



npm
 
ERR!
 A complete log of this run can be found in:

npm
 
ERR!
 /root/.npm/_logs/2020-02-22T21_07_27_092Z-debug.log

Removing intermediate container db907a5ad47b
The command '/bin/sh -c npm run build' returned a non-zero code: 2
```

I assumed because of this:

```
/app/node_modules/decco/./ppx: line 14: /app/node_modules/decco/./ppx-linux.exe: not found
```
...that related with *native* dependencies. What is that?

**Comment by r17x** [comment]
@faultable  decco it's library with native dependencies for make `decode/encode` function automagically for `type`.

File: web/src/API.re [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  type t =
+    | URLs;
+
+  let unwrap =
+    fun
+    | URLs => "/api/url/list";
+};
+
+module URLs = {
+  [@decco]
```
Comment (r17x) [inline web/src/API.re:21]: This is an example `how to use decco`. 
basically, reason don't have a type for JSON, so we need `decode/encode` from json to reason type.

File: web/src/API.re [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    slug: string,
+    destination: string,
+    ts: string,
+  };
+
+  [@decco]
+  type t = {data: option(list(url))};
+
+  let decode: Js.Json.t => Belt.Result.t(t, string) =
+    json => json->t_decode |> Relude.Result.mapError(_ => "Decode failed");
```
Comment (r17x) [inline web/src/API.re:34]: can you see `t_decode` (we never define or write `t_decode`), but this generated by `[@decco]`

**Comment by r17x** [comment]
@faultable  I think this is an environmental `issue`.  this is work correctly in my local computer 
image


anyway, what's the version OS & `bs-platform` in `ci/dockerbuild` ?

**Comment by faultables** [comment]
> @faultable I think this is an environmental `issue`. this is work correctly in my local computer
> image
> 
> anyway, what's the version OS & `bs-platform` in `ci/dockerbuild` ?

as same as with in `package.json` and `package-lock.json`

**Review by faultables — APPROVED** [review]

**Comment by faultables** [comment]
@ri7nz udah gw invite ke docker hub org, harusnya udah bisa liat build statusnya

**Pull request closed without merging**